### PR TITLE
duplicate some ops to enable more fusion opportunities and reduce memory footprint

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_duplicate_computation_for_fusion.cc [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+//    pattern #1: read buffer `%0`, `%5`, writer buffer `%7`
+//
+//  As you can see from the above analysis, we increase one read of `%0`, which
+//  is a scalar buffer, and reduce one read and one write of `%2`, which is
+//  larger than `%0`.
+//
+//  Modern NN networks are usually composed of multiple similar layers. Thus the
+//  above patterns are very common especailly when we enable shape constraint ir
+//  optimization (if enabled, we will do shape prpagation egaerly, and may
+//  further enable cross layer CSE, which in turn increases the change of the
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/disc_duplicate_computation_for_fusion.cc:118]: change -> chance?

**Review by qiuxiafei — APPROVED** [review]

File: tao_compiler/mlir/disc/transforms/disc_duplicate_computation_for_fusion.cc (hunk 1 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/disc_duplicate_computation_for_fusion.cc]: Done, thanks.

File: tao_compiler/mlir/disc/transforms/disc_passes.td [hunk 2]
```diff
@@ -406,4 +406,15 @@ def DiscStripShapeConstraintOpsPass : Pass<"disc-strip-shape-constraint-ops", "M
...
   let constructor = "createForLoopUnrollInterleavePass()";
+}
+
+def DiscDuplicateComputationForFusionPass : Pass<"disc-duplicate-computation-for-fusion", "mlir::func::FuncOp"> {
+  let summary = "Duplicate some computations in order to enable more fusion opportunity and remove memory footprint.";
+  let constructor = "createDiscDuplicateComputationForFusionPass()";
+  let options = [
+    Option<"gpu_enabled_", "gpu-enabled", "bool",
+           /*default=*/"true", "whether gpu is available.">,
+    Option<"fusion_strategy_", "fusion-strategy", "std::string",
```
Comment (Yancey1989) [inline tao_compiler/mlir/disc/transforms/disc_passes.td:417]: Do we actually need the `fusion-strategy ` option? If it is always  `base`, we can remove it in this PR, and add this option if it's required.

**Comment by Yancey1989** [comment]
BTW, any benchmark data on a model about this pass?

**Comment by wyzero** [comment]
> BTW, any benchmark data on a model about this pass?

In my test case, it reduces around ~1.5ms (e2e is ~6.5ms). I haven't test this feature on other models. Thus I do not enable this feature by default (guarded by shape-constraint-ir flag). I'll evaluate `shape-constraint-ir` on more models in next month, and then make a decision if it's ready to enable this feature by default.

File: tao_compiler/mlir/disc/transforms/disc_passes.td (hunk 2 above)
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/disc_passes.td:417]: I prefer to leave the option here even though we do not use it now actually. It's better if we make use of such config. Currently implementation is just a conservative strategy.  Furthermore, we only duplicate scalar-bcast pattern in this PR, while eventually we will need a general "duplicate fusion" pass like XLA.

**Pull request merged**